### PR TITLE
Added `Joint_Handmade` and `MarginalDistribution` in stats 

### PR DESCRIPTION
**Pull request by akash9712**
#### Brief description of what is fixed or changed
Added `JointRV`, which allows the user to create a joint random variable
with each of its component varying over the real line. Also added
`JointDistributionHandmade` for this purpose. 
Added `MarginalDistribution` to allow random variables with components marginalized out. This also adds the ability to handle compound random variables by marginalizing over latent distributions, with or without a given condition.

#### Other comments
Same code as #14951 is added, which has to be closed due to some extra commits showing up.

#### Release Notes

* stats
    * Add `JointRV` to allow user to create Joint random variables
    * Add `MarginalDistribution`, `CompoundDistribution`
    * Ability to handle compound distributions

### CODE REVIEW

**Comment by Upabjojr** [comment]
When I look at
```python
In [11]:     x = Symbol('x')
    ...:     l = Symbol('l', positive=True)
    ...:     rate = Beta(l, 2, 3)
    ...:     X = Poisson(x, rate)

In [12]: X
Out[12]: x

In [13]: X.pspace
Out[13]: JointPSpace(x, MarginalDistribution(PoissonDistribution(l), (x,)))
```
I mean, `PoissonDistribution(l)` where `l` is a random symbol is a bit weird, because the distribution is not a single continuous distribution as the class structure would lead to think:
```python
In [23]: pd
Out[23]: PoissonDistribution(l)

In [24]: type(pd)
Out[24]: sympy.stats.drv_types.PoissonDistribution

In [25]: type(pd).__mro__
Out[25]: 
(sympy.stats.drv_types.PoissonDistribution,
 sympy.stats.drv.SingleDiscreteDistribution,
 sympy.core.basic.Basic,
 sympy.stats.rv.NamedArgsMixin,
 object)
```

**Comment by Upabjojr** [comment]
Can we represent the joint distribution of a compound random variable?

**Comment by Upabjojr** [comment]
I guess the best way to represent a compound distribution is by having a `CompoundDistribution` object.

File: sympy/stats/joint_rv.py [hunk 1]
```diff
@@ -46,26 +57,49 @@ def value(self):
 
     @property
     def component_count(self):
-        return len(self.distribution.set.args)
+        _set = self.distribution.set
+        if isinstance(_set, ProductSet):
```
Comment (ArifAhmed1995) [inline sympy/stats/joint_rv.py]: You could use ternary operator here.
`return len(_set.args) if isinstance(_set, ProductSet) else 1`
Doubt would affect execution time though.

File: sympy/stats/joint_rv.py [hunk 2]
```diff
@@ -162,3 +207,115 @@ def marginal_distribution(rv, *indices):
...
+
+class CompoundDistribution(Basic, NamedArgsMixin):
+    """
+    Represents a compound probability distribution.
+
+    Constructed using a single probability distribution with a parameter
+    distributed according to some given distribution.
+    """
+    def __new__(cls, dist):
+        assert isinstance(dist, (ContinuousDistribution, DiscreteDistribution))
```
Comment (ArifAhmed1995) [inline sympy/stats/joint_rv.py]: Might be beneficial for end-user to get a a better explicit error message if instantiation is attempted to be done from an illegal distribution.
Such as:
```
if not isinstance(dist, (ContinuousDistribution, DiscreteDistribution)):
    raise ValueError('CompoundDistribution can only be initialized from ContinuousDistribution or DiscreteDistribution')
```
or something similar.

File: sympy/stats/joint_rv.py [hunk 3]
```diff
@@ -162,3 +207,115 @@ def marginal_distribution(rv, *indices):
...
+        return self.args[0].set
+
+    def __call__(self, *args):
+        return self.pdf(args)
+
+
+class MarginalDistribution(Basic):
+
+    def __new__(cls,dist, rvs):
+        assert all([isinstance(rv, (Indexed, RandomSymbol))] for rv in rvs)
```
Comment (ArifAhmed1995) [inline sympy/stats/joint_rv.py]: Same as discussed earlier. Better for end-user to get a proper error message.

File: sympy/stats/tests/test_joint_rv.py [hunk 4]
```diff
@@ -58,3 +59,12 @@ def test_JointPSpace_margial_distribution():
     assert marginal_distribution(T, T[1])(x) == sqrt(2)*(x**2 + 2)/(
         8*polar_lift(x**2/2 + 1)**(5/2))
     assert integrate(marginal_distribution(T, 1)(x), (x, -oo, oo)) == 1
+
+def test_JointRV():
+    from sympy.stats.joint_rv import JointDistributionHandmade
+    x1, x2 = (Indexed('x', i) for i in (1, 2))
+    pdf = exp(-x1**2/2 + x1 - x2**2/2 - S(1)/2)/(2*pi)
+    X = JointRV('x', pdf)
+    density(X)(1, 2) == exp(-1)/(2*pi)
```
[the remark this answers is not in the document]
Comment (akash9712) [reply sympy/stats/tests/test_joint_rv.py]: Thanks, that was an incorrect test. Fixed it now.

File: sympy/stats/tests/test_mix.py [hunk 5]
```diff
@@ -1,15 +1,28 @@
...
     assert density(X, Eq(rate, rate.symbol)) == PoissonDistribution(l)
+    N1 = Normal('N1', 0, 1)
+    N2 = Normal('N2', N1, 2)
+    assert density(N2)(0).doit() == sqrt(10)/(10*sqrt(pi))
+    assert simplify(density(N2, Eq(N1, 1))(x)) == \
+        sqrt(2)*exp(-(x - 1)**2/8)/(4*sqrt(pi))
+
+def test_compound_distribution():
+    Y = Poisson('Y', 1)
+    Z = Poisson('Z', Y)
```
Comment (Upabjojr) [inline sympy/stats/tests/test_mix.py]: I tried `Z.pspace.distribution.pdf(x)` but I get an exception.

**Pull request merged**